src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -1322,17 +1331,84 @@ static inline int MPIDI_NM_mpi_fetch_and_op(const void *origin_addr,
         goto fn_exit;
     }
 
-    /*  This can be optimized by directly calling the fi directly
-     *  and avoiding all the datatype processing of the full
-     *  MPID_Get_accumulate
-     */
-    mpi_errno = MPIDI_OFI_do_get_accumulate(origin_addr, 1, datatype,
-                                            result_addr, 1, datatype,
-                                            target_rank, target_disp, 1, datatype, op, win, NULL);
+    MPIDI_CH4U_RMA_OP_CHECK_SYNC(target_rank, win);
+    if (unlikely(target_rank == MPI_PROC_NULL))
+        goto fn_exit;
+
+    offset = target_disp * MPIDI_OFI_winfo_disp_unit(win, target_rank);
+
+    MPIDI_Datatype_check_size_lb(datatype, 1, bytes, true_lb);
+    if (bytes == 0)
+        goto fn_exit;
+
+    buffer = (char *) origin_addr + true_lb;
+    rbuffer = (char *) result_addr + true_lb;
+    tbuffer = (void *) (MPIDI_OFI_winfo_base(win, target_rank) + offset);
+
+    MPIDI_OFI_query_acc_atomic_support(datatype, MPIDI_OFI_QUERY_FETCH_ATOMIC_COUNT,
+                                       op, win, &fi_dt, &fi_op, &max_count, &dt_size);
+    if (max_count == 0)
+        goto am_fallback;
+
+    max_size = MPIDI_OFI_check_acc_order_size(win, dt_size);
+    /* It's impossible to transfer data if buffer size is smaller than basic datatype size.
+     *  TODO: we assume all processes should use the same max_size and dt_size, true ? */
+    if (max_size < dt_size)
+        goto am_fallback;
+
+    if (unlikely(op == MPI_NO_OP)) {
+        /* Fetch_and_op is READ and WRITE, except NO_OP (it is READ only). */
+        MPIDI_CH4U_wait_am_acc(win, target_rank, MPIDI_CH4I_ACCU_ORDER_RAW);
+    } else {
+        MPIDI_CH4U_wait_am_acc(win, target_rank,
+                               (MPIDI_CH4I_ACCU_ORDER_RAW | MPIDI_CH4I_ACCU_ORDER_WAR |
+                                MPIDI_CH4I_ACCU_ORDER_WAW));
+    }
+
+    originv.addr = (void *) buffer;
+    originv.count = 1;
+    resultv.addr = (void *) rbuffer;
+    resultv.count = 1;
+    targetv.addr = (uint64_t) tbuffer;
+    targetv.count = 1;
+    targetv.key = MPIDI_OFI_winfo_mr_key(win, target_rank);
+
+    MPIDI_OFI_ASSERT_IOVEC_ALIGN(&originv);
+    msg.msg_iov = &originv;
+    msg.desc = NULL;
+    msg.iov_count = 1;
+    msg.addr = MPIDI_OFI_comm_to_phys(win->comm_ptr, target_rank);
```

Comment:
You might be able to use `addr` here to avoid double translation?
Actually this topic is not limited to this function, so maybe we could address it in a separate PR to change every RMA function in OFI.

Reply:
Yeah, let's do it in a sperate PR.